### ISSUE TITLE
[QUESTION] longhorn-1.2.3 RWX very low performance on many small files

### ISSUE THREAD

**Issue by warmanton** [longhorn/longhorn#3664]
## Question
> Suggest to use https://github.com/longhorn/longhorn/discussions to ask questions.

## Environment

 - Longhorn version: 1.2.3
 - Kubernetes version: v1.23.4
 - Node config
   - OS type and version: Ubuntu 20.04.3 LTS
   - CPU per node: 6
   - Memory per node:
   - Disk type: SSD
   - Network bandwidth and latency between the nodes: 1Gbit and 0..25ms (depending if nodes in one DC or not)
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal

## Additional context

Volume in RWX mode gives very low performance on many small files. I.e. when cp one big file from pod to mount point directory then I got expected performance but when I try to extract tar archive with a lot of small files (multiple git --bare repos) then I got 3-4 files per second no matter of files size. `tar xf /tmp/archive-2022-02-25.tar -C /data -v -v` shows 3-4 lines per second. /data is mounted in RWX mode.

I suspect that this is related to NFS mount options (when RWX is requested then longhorn uses NFS mount) but I cannot find any mount options to try to tune.

Please give an advice about how to fix this or point to right place in documentation.

**Comment by derekbit** [comment]
What's the average size of small files?

**Comment by warmanton** [comment]
> What's the average size of small files?

from bytes to kBytes. For example 41 byte , 200 kB. The content is like usual `.git` directory files. I.e. I untar about 30 git repos and it already took 2 hours. It is not depending on size but from number of files.

**Comment by warmanton** [comment]
Example of testing:
Testing script is
```
#!/bin/ash

for i in $(seq -w 1000)
do
  echo "some text" > "file_${i}.txt"
done

```
When running this script on RWO mount point it creaes 1k files very very fast but when running this script in pod with RWX mountoint it creates 1k files in 2m46s :-(. This is on the same k8s cluster and only volume type is RWO vs RWX.

Volumes have difference only in access mode RWO and RWX. RWX uses NFS. The most probable place of this issue is NFS server/client or mount options.

**Comment by derekbit** [comment]
Thanks for your information. I will investigate it.
Could you please help check if the NFS mountpoint has `sync` flag?

**Comment by warmanton** [comment]
Yes I will try. Just tell me how to check this or how to view all mount options used in RWX volume. Also how to change it. I did not find by my self any place where mount options are specified.

**Comment by derekbit** [comment]
You can check `mount | grep nfs` on the node where the workload using the rwx volume is running on.
Here is the result I got and no `sync` flag
Let me dig into the performance issue more. 
```
10.43.228.22:/pvc-6f42b28e-6971-474a-a87c-9d5e1ac44f20 on /var/lib/kubelet/pods/9ca77865-8459-47c9-a005-ce58251dae08/volumes/kubernetes.io~csi/pvc-6f42b28e-6971-474a-a87c-9d5e1ac44f20/mount type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,soft,noresvport,proto=tcp,timeo=30,retrans=3,sec=sys,clientaddr=10.20.90.51,local_lock=none,addr=10.43.228.22)
```

**Comment by warmanton** [comment]
`10.43.231.202:/pvc-782e8c22-42b9-467f-9492-a8760df0af61 on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-782e8c22-42b9-467f-9492-a8760df0af61/globalmount type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,soft,noresvport,proto=tcp,timeo=30,retrans=3,sec=sys,clientaddr=10.0.2.157,local_lock=none,addr=10.43.231.202)`

See no `sync` flag here. Thanks for pointing me to the right place where I can see mount options. Previously I restarted pod, forgt about this and it went to another node, I did not find mount point on the node.

**Comment by warmanton** [comment]
B.t.w. we have two bare-metal nodes and k8s inside it (multiple libvirt workers and cnrol-plane ...). baremetal nodes have network delay about 25 ms. I think this is not too much for k8s functionality. Also I found some tests of longhorn where 100ms was fine. RWO works well, RWX has issue with many files - can the network delay be the reason of such behavior ?

**Comment by derekbit** [comment]
Well, I don't think the network delay is the root cause.
One of the possible reason might me the userspace NFS used for rwx volume.
As for the performance comparison between a RWO volume and a RWX volume, I'm not quite sure now.

**Comment by derekbit** [comment]
We have performance benchmark between RWO and RWX volume for your reference.
https://github.com/longhorn/longhorn/issues/2986#issuecomment-923404663
The random IOPs of RWX/RWO is about 12-17%.

We will benchmark the small and large files performance later.

**Comment by venomone** [comment]
Similar issue here, I have 4 worker nodes all connected with 2,5 Gbit/s and the RWX share is slow as hell, I'm also working with lots of small files.

**Comment by thecodeassassin** [comment]
I did some testing and I can confirm that RWX is many orders of magnitude slower than RWO.

I am using Hetzner's AX41-nvme servers, so both storage speed and networking speed is not an issue here.

```
RWO:

/volume-rwo # time -p tar -xf VirtualBox-6.1.38.tar.bz2
real 20.42
user 19.66
sys 4.10

RWX:

/volume-rwx # time -p tar -xf VirtualBox-6.1.38.tar.bz2
real 170.77
user 26.60
sys 15.20
```

**Comment by james-munson** [comment]
This is intrinsic to NFS itself, and not just the implementation Longhorn is using.  I've seen it on my home network when copying directories to a local NAS storage box.

Whether a file is small or large, there is a fixed overhead to make the network round trip, to write the directory structure and the file attributes.  As one site points out, `For metadata-intensive workloads NFS is often quite slow, due to the need for cache revalidation and commit-on-close.`

However, there is one thing that has changed in recent Longhorn code, and that is the ability to override and tune the RWX volume mount options in storageclass parameters.  See issue https://github.com/longhorn/longhorn/issues/6655.

Closing this as resolved.

**Comment by derekbit** [comment]
Hello @james-munson
I'm thinking if we can compare the performance of large and small files using fio (kbench). Then, users can understand the difference before using RWX volumes. WDYT?

**Comment by james-munson** [comment]
See also https://github.com/longhorn/longhorn/discussions/6964
